Features requested for warning

Ianardo-DiCaprio opened the pull request:
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
This adds commands requested in #2800. It allows to toggle warns being sent to DM's, allows to toggle warns being sent to a channel on a server and has a command to set the channel the warns get sent to, if no channel is set it gets sent to the channel the [p]warn command was used in.

Ianardo-DiCaprio committed:
New Features For Warning

Ianardo-DiCaprio committed:
blacked

Ianardo-DiCaprio committed:
-l 99

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -24,7 +25,14 @@
+        "actions": [],
+        "reasons": {},
+        "allow_custom_reasons": False,
+        "toggle_dm": True,
```
IMO, this should default to False so that guild owners/admins can choose whether or not for this to be enabled.

Jackenmen replied:
I agree, though warnings cog was acting as if it was set to `True`, so this would change that behavior and it should be included in changelog, if that's going to be changed to `False` as default.

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +84,49 @@ def __init__(self, bot: Red):
+        guild = ctx.guild
+        toggle = await self.config.guild(guild).toggle_dm()
+        await self.config.guild(guild).toggle_dm.set(not toggle)
+        toggle = await self.config.guild(ctx.guild).toggle_dm()
```
You have the toggle variable already, so you should probably use `not toggle` instead of making another Config call.

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +84,49 @@ def __init__(self, bot: Red):
+
+    @warningset.command()
+    @commands.guild_only()
+    async def warnchannel(self, ctx: commands.Context, channel=None):
```
You can use type casting for this to make it require a `discord.TextChannel`, so then you don't have to convert it later.

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +84,49 @@ def __init__(self, bot: Red):
+        guild = ctx.guild
+        toggle = await self.config.guild(guild).toggle_channel()
+        await self.config.guild(guild).toggle_channel.set(not toggle)
+        toggle = await self.config.guild(ctx.guild).toggle_channel()
```
Same thing with above, you can use `not toggle` instead of calling Config again.

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +353,41 @@ def __init__(self, bot: Red):
-            pass
+            warn_channel = await self.config.guild(ctx.guild).warn_channel()
+            if warn_channel == None:
+                channel == ctx.channel
```
Is this supposed to be a `=`?

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +353,41 @@ def __init__(self, bot: Red):
+            if warn_channel == None:
+                channel == ctx.channel
+            else:
+                channel == warn_channel
```
Is this supposed to be a `=`?

NeuroAssassin commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +353,41 @@ def __init__(self, bot: Red):
+                    _("{user} has been warned.").format(user=user.mention), embed=em
+                )
+        if dm and toggle_channel == False:
+            ctx.send("You have warned the user but you have DM and channel warns toggled off.")
```
Forgot an await here.

NeuroAssassin reviewed (changes requested):
I've noticed a couple things while taking a quick look at your PR.  I've made a list of them below.

Ianardo-DiCaprio committed:
Review changes.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
+            if warn_channel == None:
+                channel = ctx.channel
+            else:
+                channel == warn_channel
```
Second `==` still not changed to `=`, so I'm mentioning in case you didn't see.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
+                await channel.send(
+                    _("{user} has been warned.").format(user=user.mention), embed=em
+                )
+        if dm and toggle_channel == False:
```
I would remove this whole section and go back to how it was handled in this line, because there's no need to inform the user each time that DM and channel warns are toggled off:
https://github.com/Cog-Creators/Red-DiscordBot/pull/2814/files#diff-bd66023271557007376afcf648f3b3ffL341

Ianardo-DiCaprio replied:
I feel like it should as there is quite a big chance that not every moderator in a server would know exactly how the bot is set up and may not be familiar with the bot and what it means if it says nothing after a warn has been given.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
         await warning_points_add_check(self.config, ctx, user, current_point_count)
         try:
+            dm = await self.config.guild(ctx.guild).toggle_dm()
+            if dm == True:
```
You can change it to just:
```suggestion
            if dm:
```

jack1142 commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
+                    embed=em,
+                )
+                await ctx.send(_("User {user} has been warned in their DM's.").format(user=user))
+        except discord.HTTPException:
```
This `try` block is too wide, currently it may catch `discord.HTTPException` for `ctx.send()`

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
-                ),
-                embed=em,
+            warn_channel = await self.config.guild(ctx.guild).warn_channel()
+            if warn_channel == None:
```
```suggestion
            if warn_channel is None:
```
You should just use identity check, because it's singleton.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +346,43 @@ def __init__(self, bot: Red):
+                "The warning couldn't be sent to the user because the user has DM's from server members turned off."
+            )
+        toggle_channel = await self.config.guild(ctx.guild).toggle_channel()
+        if toggle_channel == True:
```
You can change it to just:
```suggestion
        if toggle_channel:
```

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +83,43 @@ def __init__(self, bot: Red):
+        toggle = await self.config.guild(guild).toggle_dm()
+        await self.config.guild(guild).toggle_dm.set(not toggle)
+        toggle = not toggle
+        if toggle == True:
```
You can change it to just:
```suggestion
        if toggle:
```
Or even not assign line above and just do `if not toggle:`

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +83,43 @@ def __init__(self, bot: Red):
+    async def warnchannel(self, ctx: commands.Context, channel: discord.TextChannel):
+        """Set the channel that warns get set to"""
+        guild = ctx.guild
+        try:
```
This `try` block doesn't seem to be needed anymore.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +83,43 @@ def __init__(self, bot: Red):
+        toggle = await self.config.guild(guild).toggle_channel()
+        await self.config.guild(guild).toggle_channel.set(not toggle)
+        toggle = not toggle
+        if toggle == True:
```
You can change it to just:
```suggestion
        if toggle:
```
Or even not assign line above and just do `if not toggle:`

Ianardo-DiCaprio committed:
Review changes.

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +342,39 @@ def __init__(self, bot: Red):
+                channel = ctx.channel
+            else:
+                channel = warn_channel
+                await channel.send(
```
I think this could work nicely if you would unindent this `channel.send()` once, so that it gets sent, when warn channel is not set too (see my proposed addition in comment to `[p]warningset warnchannel` for additional context)

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +342,39 @@ def __init__(self, bot: Red):
+            )
+        await ctx.send(_("User {user} has been warned in their DM's.").format(user=user))
+        toggle_channel = await self.config.guild(ctx.guild).toggle_channel()
+        if toggle_channel:
```
You should add some response if toggle is set to disabled (as in add `else` block after this `if`), something like there was before, for example:
```py
else:
     await ctx.send(_("User {user} has been warned.").format(user=user))
```

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -301,19 +342,39 @@ def __init__(self, bot: Red):
+            await ctx.send(
+                "The warning couldn't be sent to the user because the user has DM's from server members turned off."
+            )
+        await ctx.send(_("User {user} has been warned in their DM's.").format(user=user))
```
In my opinion, this message is unnecessary (as user that used this command should already know what settings he has), but if you want to keep it, put this in `else` block and check if DMs are enabled (cause otherwise, this message will be sent even if the warning couldn't be sent to the user or when sending DMs is disabled in cog's settings):
```suggestion
        else:
            if dm:
                await ctx.send(_("User {user} has been warned in their DMs.").format(user=user))
```

Jackenmen commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -76,6 +83,39 @@ def __init__(self, bot: Red):
+
+    @warningset.command()
+    @commands.guild_only()
+    async def warnchannel(self, ctx: commands.Context, channel: discord.TextChannel):
```
I think that it should be allowed to reset this setting (as in, set it to `None` when channel is not provided), so that bot would use `ctx.channel` in `[p]warn` command, when `togglechannel` is enabled and this is not set.

Ianardo-DiCaprio committed:
Review changes + fix

Made `[p]warningset warnchannel` set the channel to None again if no channel is said.
Made "("{user} has been warned."" send if it is being sent in channels in the channel the command is used or the set channel if one is set.
Removed the message saying that the user has been warned in their DM's
Added ctx.tick() to the warn command when the command has completed instead.
Fixed sending the warn in the channel if the channel is set.

Ianardo-DiCaprio committed:
black -l 99

Ianardo-DiCaprio committed:
Removed unneeded import

Ianardo-DiCaprio committed:
Why did it unblack itself from removing 1 thing...

Jackenmen commented:
I forgot, you should use `_("string")` format for strings, so that they will be available for translations. For most of those that miss it, it should be as simple as adding `_(` before and `)` after the string, but in the place where you used f-string, you will have to change it to regular string and use `.format()`, like in other places where translation strings are used.

Ianardo-DiCaprio committed:
_ translator

Ianardo-DiCaprio committed:
black annoys me

Ianardo-DiCaprio committed:
Merge pull request #1 from Cog-Creators/V3/develop

update

Ianardo-DiCaprio committed:
Changelog
